node/src/components/chain_synchronizer/operations.rs

```diff
@@ -1117,9 +1120,18 @@ async fn fetch_to_genesis(trusted_block: &Block, ctx: &ChainSyncContext<'_>) ->
         "starting fetch to genesis",
     );
 
-    fetch_headers_till_genesis(trusted_block, ctx).await?;
-
-    fetch_blocks_and_state_since_genesis(ctx).await?;
+    let maybe_latest_switch_block_header = fetch_headers_till_genesis(trusted_block, ctx).await?;
+    if let Some(latest_switch_block_header) = maybe_latest_switch_block_header {
+        if let Some(validator_weights) = latest_switch_block_header.next_era_validator_weights() {
+            info!(switch_block_height = %latest_switch_block_header.height(), "using validator weights");
+            fetch_blocks_and_state_and_finality_signatures_since_genesis(validator_weights, ctx)
```

Comment:
I'm afraid this won't be correct. `validator_weights` here is taken from `latest_switch_block_header.next_era_validator_weights()` which on its own is said to be `heighest switch block` but we need the _lowest_ one - and we would start from Genesis validators because we start from Genesis and move forward - in order to validate signatures in era `N` we need validators taken from the switch block of era `N-1`.

Reply:
The correct logic has been implemented in b1172ac330b0abd8ef94e017ec280fe856d022b6